pkg/tracing/trace.go

```diff
@@ -0,0 +1,118 @@
+)
+
+var msmInstance MigrationSpanMap
+var createMsmMapOnce sync.Once
+
+// MigrationSpanMap provides a map between MigMigration UID and associated Jaeger span.
+// This is required so that all controllers can attach child spans to the correct
+// parent migmigration span for unified tracing of work done during migrations.
+type MigrationSpanMap struct {
+	mutex              sync.RWMutex
+	migrationUIDToSpan map[string]opentracing.Span
+}
+
+// InitJaeger returns an instance of Jaeger Tracer that samples 100% of traces and logs all spans to stdout.
+func InitJaeger(service string) (opentracing.Tracer, io.Closer) {
+	cfg := &config.Configuration{
+		ServiceName: service,
+		Sampler: &config.SamplerConfig{
+			Type:  "const",
+			Param: 1,
+		},
+		Reporter: &config.ReporterConfig{
+			LogSpans: true,
+		},
+	}
+	tracer, closer, err := cfg.NewTracer(config.Logger(jaeger.StdLogger))
+	if err != nil {
+		panic(fmt.Sprintf("ERROR: cannot init Jaeger: %v\n", err))
+	}
+	return tracer, closer
+}
+
+// SetSpanForMigrationUID sets the parent jaeger span for a migration
+func SetSpanForMigrationUID(migrationUID string, span opentracing.Span) {
+	// Init map if needed
+	createMsmMapOnce.Do(func() {
+		msmInstance = MigrationSpanMap{}
+		msmInstance.migrationUIDToSpan = make(map[string]opentracing.Span)
+	})
+	msmInstance.mutex.RLock()
+	defer msmInstance.mutex.RUnlock()
+
+	msmInstance.migrationUIDToSpan[migrationUID] = span
+	return
+}
+
+// RemoveSpanForMigrationUID removes a span from the span map once migration is complete.
+func RemoveSpanForMigrationUID(migrationUID string) {
+	// Init map if needed
+	createMsmMapOnce.Do(func() {
+		msmInstance = MigrationSpanMap{}
+		msmInstance.migrationUIDToSpan = make(map[string]opentracing.Span)
+	})
+	msmInstance.mutex.RLock()
+	defer msmInstance.mutex.RUnlock()
+
+	delete(msmInstance.migrationUIDToSpan, migrationUID)
+	return
+}
+
+// GetSpanForMigrationUID returns the parent jaeger span for a migration
+func GetSpanForMigrationUID(migrationUID string) opentracing.Span {
+	// Init map if needed
+	createMsmMapOnce.Do(func() {
+		msmInstance = MigrationSpanMap{}
+		msmInstance.migrationUIDToSpan = make(map[string]opentracing.Span)
+	})
+	msmInstance.mutex.RLock()
+	defer msmInstance.mutex.RUnlock()
+
+	migrationSpan, ok := msmInstance.migrationUIDToSpan[migrationUID]
+	if !ok {
+		return nil
+	}
+	return migrationSpan
+}
+
+// CloseMigrationSpan closes out the parent Migration Span safely
+func CloseMigrationSpan(migrationUID string) {
+	migrationSpan := GetSpanForMigrationUID(migrationUID)
+	if migrationSpan != nil {
+		// Immediately remove span from map so other controllers stop adding child spans
+		RemoveSpanForMigrationUID(migrationUID)
+		// Wait 5 minutes before terminating span, since other controllers writing to span
+		// post-close will result in undefined jaeger behavior (e.g. broken statistics page)
+		go func(migrationSpan opentracing.Span) {
+			time.Sleep(5 * time.Minute)
```

Comment:
Is there a better way to note that other things are potentially writing to the span rather than relay on a timeout?

Reply:
I thought about using a semaphore here or something similar, I think that would be a more robust solution. 

This was a sort of quick and easy hack that I think will be effective as long as no controller has a reconcile extending past 5 minutes. In general our reconciles are much shorter than this, with the exception of DirectImageMigration related things.

I actually wasn't sure why jaeger data corruption was happening, but I was able to confirm that this fixed the problem by making sure that the top-level migration span isn't written to post-close.

Reply:
If you feel this is important to fix now I can definitely do it before merge.

Reply:
I will plan to fix this if we see any problems with the current approach in the future, or if we plan to turn Jaeger tracing on outside of dev environments. Filed an issue.

Reply:
https://github.com/konveyor/mig-controller/issues/1023